File: app/src/main/java/com/ibashkimi/lockscheduler/model/TimeCondition.java
package com.ibashkimi.lockscheduler.model;

public class TimeCondition extends Condition {

    private boolean[] daysActive;

    private Time startTime;

    private Time endTime;

    public TimeCondition() {
        this(new boolean[]{true, true, true, true, true, true, true});
    }

    public TimeCondition(boolean[] daysActive) {
        this(daysActive, new Time(0, 0), new Time(0, 0));
    }

    public TimeCondition(boolean[] daysActive, Time startTime, Time endTime) {
        super(Type.TIME);
        this.daysActive = daysActive;
        this.startTime = startTime;
        this.endTime = endTime;
    }

    public boolean[] getDaysActive() {
        return daysActive;
    }

    public Time getStartTime() {
        return startTime;
    }

    public Time getEndTime() {
        return endTime;
    }

    public void setStartTime(Time startTime) {
        this.startTime = startTime;
    }

    public void setEndTime(Time endTime) {
        this.endTime = endTime;
    }

    public void setDaysActive(boolean[] daysActive) {
        this.daysActive = daysActive;
    }

    @Override
    public String toString() {
        return "TimeCondition[" + daysActive[0] + daysActive[1] + daysActive[2] + daysActive[3] + daysActive[4] + daysActive[5] + daysActive[6] + ", startTime=" + startTime + ", endTime=" + endTime + "]";
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof TimeCondition))
            return false;
        TimeCondition condition = (TimeCondition) obj;
        for (int i = 0; i < 7; i++) {
            if (condition.getDaysActive()[i] != daysActive[i])
                return false;
        }
        return condition.startTime.equals(startTime) && condition.endTime.equals(endTime);
    }
}

File: app/src/main/java/com/ibashkimi/lockscheduler/model/condition/TimeCondition.kt
package com.ibashkimi.lockscheduler.model.condition

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize

@Parcelize
data class TimeCondition(
    var daysActive: DaysOfWeek = DaysOfWeek(),
    var startTime: Time = Time(0, 0),
    var endTime: Time = Time(0, 0),
    override var isTriggered: Boolean = false
) : Condition(Type.TIME, isTriggered), Parcelable

@Parcelize
data class DaysOfWeek(
    var monday: Boolean = true,
    var tuesday: Boolean = true,
    var wednesday: Boolean = true,
    var thursday: Boolean = true,
    var friday: Boolean = true,
    var saturday: Boolean = true,
    var sunday: Boolean = true
) : Parcelable {

    operator fun get(index: Int): Boolean {
        return when (index) {
            0 -> monday
            1 -> tuesday
            2 -> wednesday
            3 -> thursday
            4 -> friday
            5 -> saturday
            6 -> sunday
            else -> throw IndexOutOfBoundsException("Invalid index $index. Valid range: [0 : 6].")
        }
    }

    operator fun set(index: Int, value: Boolean) {
        return when (index) {
            0 -> monday = value
            1 -> tuesday = value
            2 -> wednesday = value
            3 -> thursday = value
            4 -> friday = value
            5 -> saturday = value
            6 -> sunday = value
            else -> throw IndexOutOfBoundsException("Invalid index $index. Valid range: [0 : 6].")
        }
    }

    fun activeDaysIndexes() = asBooleanArray().indices.filter { get(it) }

    fun asBooleanArray(): BooleanArray {
        return booleanArrayOf(monday, tuesday, wednesday, thursday, friday, saturday, sunday)
    }

    fun createCopy(): DaysOfWeek {
        return DaysOfWeek(monday, tuesday, wednesday, thursday, friday, saturday, sunday)
    }
}
